File: Core/src/main/java/dev/lynith/core/utils/nvg/NanoVGHelper.java
package dev.lynith.core.utils.nvg;

import dev.lynith.core.ClientStartup;
import dev.lynith.core.ui.BoundingBox;
import dev.lynith.core.ui.styles.impl.Border;
import dev.lynith.core.ui.styles.impl.Color;
import dev.lynith.core.ui.styles.impl.CornerRadius;
import dev.lynith.core.ui.styles.impl.FontStyles;
import org.lwjgl.nanovg.NVGColor;
import org.lwjgl.nanovg.NanoVG;

import static org.lwjgl.nanovg.NanoVG.*;

public class NanoVGHelper {

    public static void createFrame() {
        nvgBeginFrame(ctx(), ClientStartup.getInstance().getVersion().getRenderer().getWindowWidth(), ClientStartup.getInstance().getVersion().getRenderer().getWindowHeight(), 1);
        nvgSave(ctx());
    }

    public static void endFrame() {
        nvgRestore(ctx());
        nvgEndFrame(ctx());
    }

    public static NVGColor createColor(int r, int g, int b, int a) {
        NVGColor nvgColor = NVGColor.create();
        nvgRGBAf((float) r / 255, (float) g / 255, (float) b / 255, (float) a / 255, nvgColor);
        return nvgColor;
    }

    public static NVGColor createColor(Color color) {
        return createColor(color.getRed(), color.getGreen(), color.getBlue(), color.getAlpha());
    }

    public static void rectangle(float left, float top, float width, float height, Color color) {
        nvgBeginPath(ctx());
        nvgFillColor(ctx(), createColor(color));
        nvgRect(ctx(), left, top, width, height);
        nvgFill(ctx());
    }

    public static void rectangle(BoundingBox bounds, Color color, Border border, CornerRadius cornerRadius) {
        int radius = 0;
        if (cornerRadius != null) {
            radius = cornerRadius.getValue();
        }

        // Background
        nvgBeginPath(ctx());
        nvgFillColor(ctx(), createColor(color));
        nvgRoundedRect(ctx(), bounds.getX(), bounds.getY(), bounds.getWidth(), bounds.getHeight(), radius);
        nvgFill(ctx());

        // Border
        if (border != null && border.getValue().getFirst() > 0) {
            nvgBeginPath(ctx());
            nvgStrokeColor(ctx(), createColor(border.getValue().getSecond()));
            nvgStrokeWidth(ctx(), border.getValue().getFirst());
            nvgRoundedRect(ctx(), bounds.getX(), bounds.getY(), bounds.getWidth(), bounds.getHeight(), radius);
            nvgStroke(ctx());
        }
    }

    public static void rectangle(BoundingBox bounds, Color color, Border border) {
        rectangle(bounds, color, border, null);
    }

    public static void rectangle(BoundingBox bounds, Color color) {
        rectangle(bounds, color, null, null);
    }

    public static void circle(float centerX, float centerY, float radius, Color color) {
        nvgBeginPath(ctx());
        nvgFillColor(ctx(), createColor(color));
        nvgCircle(ctx(), centerX, centerY, radius);
        nvgFill(ctx());
    }

    public static void text(String text, float left, float top, int size, Color color, Font font) {
        nvgBeginPath(ctx());
        nvgFontFace(ctx(), font.getName());
        nvgFontSize(ctx(), size);
        nvgTextAlign(ctx(), NVG_ALIGN_LEFT | NVG_ALIGN_MIDDLE);
        nvgFillColor(ctx(), createColor(color));
        nvgText(ctx(), left, top, text);
    }

    public static void text(String text, BoundingBox box, FontStyles styles, Color color) {
        nvgBeginPath(ctx());
        nvgFontFace(ctx(), styles.getValue().getFirst());
        nvgFontSize(ctx(), styles.getValue().getSecond());
        nvgTextAlign(ctx(), NVG_ALIGN_LEFT | NVG_ALIGN_MIDDLE);
        nvgFillColor(ctx(), createColor(color));
        nvgTextBox(ctx(), box.getX(), box.getY(), box.getWidth(), text);
    }

    public static float textWidth(String text, FontStyles styles) {
        nvgBeginPath(ctx());
        nvgFontFace(ctx(), styles.getValue().getFirst());
        nvgFontSize(ctx(), styles.getValue().getSecond());
        float[] bounds = new float[4];
        System.out.println("textWidth: " + text);
        nvgTextBounds(ctx(), 0, 0, text, bounds);
        return bounds[2] - bounds[0];
    }

    public static int textWidthInt(String text, FontStyles styles) {
        return (int) textWidth(text, styles);
    }

    public static float textHeight(String text, FontStyles styles) {
        nvgBeginPath(ctx());
        nvgFontFace(ctx(), styles.getValue().getFirst());
        nvgFontSize(ctx(), styles.getValue().getSecond());
        float[] bounds = new float[4];
        System.out.println("textHeight: " + text);
        nvgTextBounds(ctx(), 0, 0, text, bounds);
        return bounds[3] - bounds[1];
    }

    public static long ctx() {
        return ClientStartup.getInstance().getNvgContext();
    }

}

File: Core/src/main/kotlin/dev/lynith/core/ui/NanoVGHelper.kt
package dev.lynith.core.ui

import dev.lynith.core.ClientStartup
import dev.lynith.core.ui.styles.impl.Border
import dev.lynith.core.ui.styles.impl.Color
import dev.lynith.core.ui.styles.impl.CornerRadius
import dev.lynith.core.ui.styles.impl.FontStyles
import dev.lynith.core.ui.theme.ThemeManager
import dev.lynith.core.utils.nvg.Font
import dev.lynith.core.utils.nvg.FontHelper
import org.lwjgl.nanovg.NVGColor
import org.lwjgl.nanovg.NanoVG.*

open class NanoVGHelper {
    companion object {
        fun createFrame() {
            nvgBeginFrame(ctx(), ClientStartup.instance.version.getRenderer().getWindowWidth().toFloat(), ClientStartup.instance.version.getRenderer().getWindowHeight().toFloat(), 1f)
            nvgSave(ctx())
        }

        fun endFrame() {
            nvgRestore(ctx())
            nvgEndFrame(ctx())
        }

        fun createColor(r: Int, g: Int, b: Int, a: Int): NVGColor {
            val nvgColor = NVGColor.create()
            nvgRGBAf(r.toFloat() / 255, g.toFloat() / 255, b.toFloat() / 255, a.toFloat() / 255, nvgColor)
            return nvgColor
        }

        fun background(color: Color = ThemeManager.instance.currentTheme.colorScheme.background) {
            rectangle(
                0f,
                0f,
                ClientStartup.instance.version.getRenderer().getWindowWidth().toFloat(),
                ClientStartup.instance.version.getRenderer().getWindowHeight().toFloat(),
                color
            )
        }

        fun createColor(color: Color): NVGColor {
            return createColor(color.red, color.green, color.blue, color.alpha)
        }

        fun rectangle(left: Float, top: Float, width: Float, height: Float, color: Color) {
            nvgBeginPath(ctx())
            nvgFillColor(ctx(), createColor(color))
            nvgRect(ctx(), left, top, width, height)
            nvgFill(ctx())
            nvgClosePath(ctx())
        }

        @JvmOverloads
        fun rectangle(bounds: BoundingBox, color: Color, border: Border? = null, cornerRadius: CornerRadius? = null) {
            var radius = 0
            if (cornerRadius != null) {
                radius = cornerRadius.value
            }

            // Background
            nvgBeginPath(ctx())
            nvgFillColor(ctx(), createColor(color))
            nvgRoundedRect(ctx(), bounds.x.toFloat(), bounds.y.toFloat(), bounds.width.toFloat(), bounds.height.toFloat(), radius.toFloat())
            nvgFill(ctx())
            nvgClosePath(ctx())

            // Border
            if (border != null && border.thickness > 0) {
                nvgBeginPath(ctx())
                nvgStrokeColor(ctx(), createColor(border.color))
                nvgStrokeWidth(ctx(), border.thickness.toFloat())
                nvgRoundedRect(ctx(), bounds.x.toFloat(), bounds.y.toFloat(), bounds.width.toFloat(), bounds.height.toFloat(), radius.toFloat())
                nvgStroke(ctx())
                nvgClosePath(ctx())
            }
        }

        fun circle(centerX: Float, centerY: Float, radius: Float, color: Color) {
            nvgBeginPath(ctx())
            nvgFillColor(ctx(), createColor(color))
            nvgCircle(ctx(), centerX, centerY, radius)
            nvgFill(ctx())
            nvgClosePath(ctx())
        }

        fun text(text: String, left: Float, top: Float, size: Float, color: Color, font: Font) {
            nvgBeginPath(ctx())
            nvgFontFace(ctx(), font.formatted())
            nvgFontSize(ctx(), size)
            nvgTextAlign(ctx(), NVG_ALIGN_LEFT or NVG_ALIGN_MIDDLE)
            nvgFillColor(ctx(), createColor(color))
            nvgText(ctx(), left, top, text)
            nvgClosePath(ctx())
        }

        fun text(text: String, bounds: BoundingBox, fontStyles: FontStyles, color: Color) {
            nvgBeginPath(ctx())
            nvgFontFace(ctx(), FontHelper.getOrDefault(fontStyles.name, fontStyles.weight).formatted())
            nvgFontSize(ctx(), fontStyles.size)
            nvgTextAlign(ctx(), textAlign(fontStyles.align) or NVG_ALIGN_BOTTOM)
            nvgFillColor(ctx(), createColor(color))
            nvgTextLetterSpacing(ctx(), fontStyles.letterSpacing)
            nvgTextBox(
                ctx(),
                bounds.x,
                bounds.y + textHeight(text, fontStyles),
                bounds.width,
                text
            )
            nvgClosePath(ctx())
        }

        fun textWidth(text: String, styles: FontStyles): Float {
            val bounds = FloatArray(4)
            nvgFontSize(ctx(), styles.size)
            nvgTextLetterSpacing(ctx(), styles.letterSpacing)
            nvgFontFace(ctx(), FontHelper.getOrDefault(styles.name, styles.weight).formatted())
            return nvgTextBounds(ctx(), 0f, 0f, text, bounds)
        }

        fun textHeight(text: String, styles: FontStyles): Float {
            val bounds = FloatArray(4)
            val font = FontHelper.getOrDefault(styles.name, styles.weight)

            nvgFontSize(ctx(), styles.size)
            nvgFontFace(ctx(), font.formatted())
            nvgTextLetterSpacing(ctx(), styles.letterSpacing)
            nvgTextLineHeight(ctx(), styles.lineHeight)
            nvgTextBoxBounds(ctx(), 0f, 0f, textWidth(text, styles), text, bounds)

            return (bounds[3] - bounds[1]) + font.offset
        }

        fun textAlign(align: Font.FontAlign): Int {
            return when (align) {
                Font.FontAlign.LEFT -> NVG_ALIGN_LEFT
                Font.FontAlign.CENTER -> NVG_ALIGN_CENTER
                Font.FontAlign.RIGHT -> NVG_ALIGN_RIGHT
            }
        }

        fun ctx() = ClientStartup.instance.nvgContext
    }
}
